### PR TITLE
rename Core.BottomType => Core.TypeofBottom

### PR DESCRIPTION
**Pull request by StefanKarpinski**

### CODE REVIEW

**Comment by tkelman** [comment]
wasn't exported, but would break JLD - https://github.com/JuliaIO/JLD.jl/blob/aced8cbe4c89712b48434e73d62ac96b582bf409/src/JLD.jl#L1039

add a non-exported deprecation?

**Comment by StefanKarpinski** [comment]
> add a non-exported deprecation?

What about just fixing the one case where it's used, i.e. JLD.

**Comment by tkelman** [comment]
Possibly used in private or non-github code, and it's not like a deprecation is hard to add?

In order to still be able to benchmark against intermediate commits JLD will also have to account for both names at different versions.

**Comment by StefanKarpinski** [comment]
Adding a deprecation for a thing that was added in this release that only one package uses is silly.

**Comment by ararslan** [comment]
Out of curiosity, what's the motivation for the change?

File: base/docs/basedocs.jl [hunk 1]
```diff
@@ -675,6 +675,14 @@ generation specialization for that field.
 ANY
 
 """
+    Core.TypeofBottom
+    Type{Union{}}
```
Comment (yuyichao) [inline base/docs/basedocs.jl]: This doesn't look right

```julia
julia> Core.BottomType == Type{Union{}}
false

julia> Core.BottomType === Type{Union{}}
false
```

Comment (JeffBezanson) [inline base/docs/basedocs.jl]: #21023 fixes that (the `==` call at least).

Comment (StefanKarpinski) [inline base/docs/basedocs.jl]: Yes, it's currently not true. Jeff and Jameson are having an ongoing debate about whether this should be true or not. Consider it aspirational.

**Comment by StefanKarpinski** [comment]
The motivation is that `BottomType` sounds like `Union{}` whereas this is the type of that, so calling it `TypeofBottom` seems considerably less confusing.

**Review by ararslan — APPROVED** [review]
macOS failure looks entirely unrelated

**Comment by StefanKarpinski** [comment]
@JeffBezanson: what do you think? Should I drop the docstring commit and then merge this?

**Comment by JeffBezanson** [comment]
I think this is ok. I would remove `Type{Union{}}` from the docstring but keep the rest, and also rename `jl_bottomtype_type` to `jl_typeofbottom_type` in src/.

**Comment by tkelman** [comment]
It's "silly" to add a single line to avoid hard-breaking a widely used package during feature freeze?

**Review by tkelman — DISMISSED** [review]
either the old name should be deprecated or a fix to JLD (that keeps the ability for the package to work at versions that used the old name) should be staged, or this will be disruptive, break benchmarking and quite a few other packages

**Comment by StefanKarpinski** [comment]
PR created for JLD. Deprecating a binding in Core is not so simple, but you're welcome to take a crack at it.

**Comment by StefanKarpinski** [comment]
> wonder why it wasn't written as typeof to begin with

Maybe because it wasn't entirely clear that `Core.BottomType == typeof(Union{})` in the first place? Whoever wrote this may have wanted `Type{Union{}}` to work, but those are currently different.

**Comment by tkelman** [comment]
Jeff wrote it - https://github.com/JuliaIO/JLD.jl/commit/1fa39297e6e6cb299ac05a71959820ced62957d2
though I think it was while the type system rewrite was still getting finished up, to get benchmarking working for it.

**Comment by StefanKarpinski** [comment]
In that case it's because he's probably used to referring to the type that way :)

**Comment by ararslan** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/21057#issuecomment-289134149) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/b7177d23986c7d771eac9a7dc240b59bb1f72b6a/de26978_vs_d6088fc/report.md). cc @jrevels

**Comment by ararslan** [comment]
Cool, looks like Nanosoldier is ready, and the AV appears unrelated. Good to merge, I think.

**Pull request merged**